### PR TITLE
Fix downscaling functionality

### PR DESCRIPTION
**Pull request by JGuetschow**
Fix issues with 0 data in downscaling functions

# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [ ] Documentation added (where applicable)
- [x] Description in a `{pr}.thing.md` file in the directory `changelog` added - see [changelog/README.md](https://github.com/pik-primap/primap2/blob/main/changelog/README.md) for details

## Description

Please provide a short description what your pull request does.

### CODE REVIEW

**Comment by JGuetschow** [comment]
I've added more tests and I've worked on the case where data is all zero only for some years. The test for dimension downscaling passes, the test for variable downscaling fails. However, the current treatment of the case is not good, as it imposes equal shares for the basket elements which influence downscaling for other years. To avoid this we have to change the structure of the code and split up the share calculation step which needs some work.

**Comment by JGuetschow** [comment]
Fixed the downscaling for mixed zero and non-zero data. Ready for review now.

**Review by mikapfl — APPROVED** [review]
Thanks, looks good. I missed the tests earlier, no idea how. The tests are pretty comprehensive and I think I understand now what the problem was. Gnarly, such a nice and easy formula failing for 0/0.

File: primap2/tests/test_downscale.py [hunk 1]
```diff
@@ -201,3 +219,136 @@ def test_downscale_timeseries(empty_ds):
...
+    expected["CO2"].loc[{"time": "2010"}] = 0 * ureg("Gg CO2 / year")
+    expected["SF6"].loc[{"time": "2010"}] = 0 * ureg("Gg SF6 / year")
+    expected["CH4"].loc[{"time": "2010"}] = 0 * ureg("Gg CH4 / year")
+
+    xr.testing.assert_identical(downscaled, expected)
+
+
+# TODO more tests:
+
+# * zero and non-zero data
```
Comment (mikapfl) [inline primap2/tests/test_downscale.py:354]: This TODO is done, right?

File: primap2/_downscale.py [hunk 2]
```diff
@@ -104,6 +111,32 @@ def downscale_timeseries(
...
+
+        any_nonzero = basket_sum.where(basket_sum != 0).notnull().any()
+
+        # treat the case where all data is zero or NaN
+        if (not basket_both_zero.isnull().all()) & (not any_nonzero):
+            unit_content = str(basket_contents_da.pint.units)
+            basket_da_squeeze = basket_da.drop(dim)
+            basket_contents_da = basket_contents_da.where(
+                (basket_da_squeeze != 0) | (basket_sum != 0), 1 * ureg(unit_content)
+            )
```
Comment (mikapfl) [inline primap2/_downscale.py:132]:
```suggestion
            units = basket_contents_da.pint.units
            basket_da_squeeze = basket_da.drop(dim)
            basket_contents_da = basket_contents_da.where(
                (basket_da_squeeze != 0) | (basket_sum != 0), 1 * units
            )
```
Is there a reason why this wouldn't work? Is the conversion-to-string-and-back necessary?

Comment (JGuetschow) [reply primap2/_downscale.py:132]: It makes no sense when looking at it. I had some problem with the unit conversion and it might just be a result of fiddling around until I had something that worked. I remember that a bit later I saw unit treatment in another function and thought that it looked more elegant than what I did. So probably it's just bad code.

File: primap2/_downscale.py [hunk 3]
```diff
@@ -104,6 +111,32 @@ def downscale_timeseries(
...
+            # so we could also remove it and say whoever switches it of has
+            # to deal with the consequences
+            error_message = generate_error_message(basket_sum_zero)
+            raise ValueError(f"pr.downscale_timeseries {error_message}")
+
+        any_nonzero = basket_sum.where(basket_sum != 0).notnull().any()
+
+        # treat the case where all data is zero or NaN
+        if (not basket_both_zero.isnull().all()) & (not any_nonzero):
+            unit_content = str(basket_contents_da.pint.units)
```
[the remark this answers is not in the document]
Comment (JGuetschow) [reply primap2/_downscale.py:128]: I had to re-read the code again to understand it, but I think your comment is describing what's done and definitely the purpose.

File: primap2/_downscale.py [hunk 4]
```diff
@@ -146,6 +184,11 @@ def downscale_timeseries(
         used to downscale the basket to its contents, which is used to fill gaps in the
         timeseries of the basket contents.
 
+        If the data to downscale contains only 0 an NaN the result will be all
+        zero timeseries. If the data is only for individual basket and basket_content
+        values the resulting downscaled data points are zero but the shares for other years
```
Comment (JGuetschow) [inline primap2/_downscale.py:189]:
```suggestion
        values, the resulting downscaled data points are zero but the shares for other years
```

File: primap2/_downscale.py [hunk 5]
```diff
@@ -146,6 +184,11 @@ def downscale_timeseries(
...
+        are not be influenced
```
Comment (JGuetschow) [inline primap2/_downscale.py:190]:
```suggestion
        are not influenced
```

File: primap2/_downscale.py [hunk 6]
```diff
@@ -265,6 +277,11 @@ def downscale_gas_timeseries(
         converted back afterwards; for both conversions the gwp conversions of the
         basket are used.
 
+        If the data to downscale contains only 0 an NaN the result will be all
+        zero timeseries. If the data is only for individual basket and basket_content
+        values the resulting downscaled data points are zero but the shares for other years
```
Comment (JGuetschow) [inline primap2/_downscale.py:282]:
```suggestion
        values, the resulting downscaled data points are zero but the shares for other years
```

File: primap2/_downscale.py [hunk 7]
```diff
@@ -265,6 +277,11 @@ def downscale_gas_timeseries(
...
+        are not be influenced
```
Comment (JGuetschow) [inline primap2/_downscale.py:283]:
```suggestion
        are not influenced
```

File: primap2/tests/test_downscale.py (hunk 1 above)
Comment (JGuetschow) [reply primap2/tests/test_downscale.py:354]: yes

File: primap2/tests/test_downscale.py [hunk 8]
```diff
@@ -201,3 +219,136 @@ def test_downscale_timeseries(empty_ds):
...
+    expected["SF6"].loc[{"time": "2020"}] = 2 * ureg("Gg SF6 / year")
+    expected["CH4"].loc[{"time": "2020"}] = 2 * ureg("Gg CH4 / year")
+    expected["CO2"].loc[{"time": "2010"}] = 0 * ureg("Gg CO2 / year")
+    expected["SF6"].loc[{"time": "2010"}] = 0 * ureg("Gg SF6 / year")
+    expected["CH4"].loc[{"time": "2010"}] = 0 * ureg("Gg CH4 / year")
+
+    xr.testing.assert_identical(downscaled, expected)
+
+
+# TODO more tests:
```
Comment (JGuetschow) [inline primap2/tests/test_downscale.py:352]:
```suggestion

```

File: primap2/tests/test_downscale.py (hunk 1 above)
Comment (JGuetschow) [inline primap2/tests/test_downscale.py:354]:
```suggestion

```
